### PR TITLE
Migration of resource binaries

### PR DESCRIPTION
**Pull request by mjs**
This PR adds APIs, updates the migration binary import/export functionality and the migrationmaster worker to support migration of resources.

The new download API uses the preexisting component architecture approach. The upload API is implemented as /migrate/resources using the same approach as is used for tools and charms migration uploads.

### QA

Migrated a resource using model between 2 controllers and confirmed that the application level resource was correctly migrated using `juju dump-model` and inspection of the DB.

### CODE REVIEW

File: api/migrationmaster/client.go [hunk 1]
```diff
@@ -152,10 +160,11 @@ func (c *Client) Prechecks() error {
 // with the API connection. The charms used by the model are also
 // returned.
 func (c *Client) Export() (migration.SerializedModel, error) {
+	var empty migration.SerializedModel
```
Comment (howbazaar) [inline api/migrationmaster/client.go]: I quite like this idiom and use it as well.

File: api/migrationmaster/client.go [hunk 2]
```diff
@@ -230,6 +259,14 @@ func (c *Client) MinionReports() (migration.MinionReports, error) {
 	return out, nil
 }
 
+func (c *Client) StreamModelLog(start time.Time) (<-chan common.LogMessage, error) {
```
Comment (howbazaar) [inline api/migrationmaster/client.go]: Docstring?

File: api/migrationmaster/client_test.go [hunk 3]
```diff
@@ -244,6 +316,22 @@ func (s *ClientSuite) TestExportError(c *gc.C) {
 	c.Assert(err, gc.ErrorMatches, "blam")
 }
 
+func (s *ClientSuite) TestOpenResource(c *gc.C) {
+	doer := &fakeDoer{
```
Comment (howbazaar) [inline api/migrationmaster/client_test.go]: Can't we have a name other than doer?

File: api/migrationmaster/client_test.go [hunk 4]
```diff
@@ -426,3 +513,29 @@ func (c fakeConnector) ConnectStream(path string, attrs url.Values) (base.Stream
...
+
+func (fakeHTTPCaller) BestFacadeVersion(string) int {
+	return 0
+}
+
+func (c fakeHTTPCaller) HTTPClient() (*httprequest.Client, error) {
+	return c.httpClient, c.err
+}
+
+type fakeDoer struct {
```
Comment (howbazaar) [inline api/migrationmaster/client_test.go]: What is this really? What is it faking?

File: api/migrationtarget/client_test.go [hunk 5]
```diff
@@ -132,6 +142,71 @@ func (s *ClientSuite) TestLatestLogTimeError(c *gc.C) {
 	s.AssertModelCall(c, stub, names.NewModelTag("fake"), "LatestLogTime", err, true)
 }
 
+func (s *ClientSuite) TestUploadCharm(c *gc.C) {
+	const charmBody = "charming"
+	curl := charm.MustParseURL("cs:~user/foo-2")
+	doer := newFakeDoer(c, params.CharmsResponse{
```
Comment (howbazaar) [inline api/migrationtarget/client_test.go]: No doer... please...

File: apiserver/migrationmaster/facade.go [hunk 6]
```diff
@@ -309,3 +310,40 @@ func addToolsVersionForMachine(machine description.Machine, usedVersions map[ver
...
+	return out
+}
+
+func resourceToSerialized(app string, desc description.Resource) params.SerializedModelResource {
+	return params.SerializedModelResource{
+		Application:         app,
+		Name:                desc.Name(),
+		ApplicationRevision: revisionToSerialized(desc.ApplicationRevision()),
+		CharmStoreRevision:  revisionToSerialized(desc.CharmStoreRevision()),
+		// TODO(menn0) - unit revisions
```
Comment (howbazaar) [inline apiserver/migrationmaster/facade.go]: Is this the next bit?

File: apiserver/resources_test.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
+// Copyright 2012, 2013 Canonical Ltd.
```
Comment (howbazaar) [inline apiserver/resources_test.go]: 2016

File: core/description/resource_test.go [hunk 8]
```diff
@@ -100,6 +100,12 @@ func (s *ResourceSuite) TestNew(c *gc.C) {
 	c.Check(csRev.Username(), gc.Equals, "")
 }
 
+func (s *ResourceSuite) TestNilRevisions(c *gc.C) {
+	r := newResource(ResourceArgs{"z"})
+	c.Check(r.ApplicationRevision(), gc.Equals, nil)
```
Comment (howbazaar) [inline core/description/resource_test.go]: should be gc.IsNil

File: core/migration/migration.go [hunk 9]
```diff
@@ -7,6 +7,7 @@ import (
 	"time"
 
 	"github.com/juju/errors"
+	"github.com/juju/juju/resource"
```
Comment (howbazaar) [inline core/migration/migration.go]: import group

File: resource/api/server/handler.go [hunk 10]
```diff
@@ -4,31 +4,44 @@
...
 	Connect func(*http.Request) (DataStore, names.Tag, error)
 
+	// HandleDownload provides the download functionality.
+	HandleDownload func(st DataStore, req *http.Request) (io.ReadCloser, int64, error)
+
 	// HandleUpload provides the upload functionality.
 	HandleUpload func(username string, st DataStore, req *http.Request) (*api.UploadResult, error)
 }
 
 // TODO(ericsnow) Can username be extracted from the request?
```
Comment (howbazaar) [inline resource/api/server/handler.go]: No one is going to answer this, can we delete it?

File: resource/api/server/handler_test.go [hunk 11]
```diff
@@ -61,23 +65,35 @@ func (s *LegacyHTTPHandlerSuite) connect(req *http.Request) (server.DataStore, n
...
+func (s *HTTPHandlerSuite) handleUpload(username string, st server.DataStore, req *http.Request) (*api.UploadResult, error) {
 	s.stub.AddCall("HandleUpload", username, st, req)
 	if err := s.stub.NextErr(); err != nil {
 		return nil, errors.Trace(err)
 	}
+	return s.uploadResult, nil
+}
 
-	return s.result, nil
+func (s *HTTPHandlerSuite) copyRequest() *http.Request {
```
Comment (howbazaar) [inline resource/api/server/handler_test.go]: nice addition

**Comment by mjs** [comment]
$$merge$$

**Pull request merged**